rust/lance/src/index/cache.rs

```diff
@@ -14,35 +14,46 @@
 
 use std::sync::Arc;
 
+use lance_index::scalar::ScalarIndex;
 use moka::sync::{Cache, ConcurrentCacheExt};
 
 use super::vector::VectorIndex;
 
 #[derive(Clone)]
 pub struct IndexCache {
-    cache: Arc<Cache<String, Arc<dyn VectorIndex>>>,
+    scalar_cache: Arc<Cache<String, Arc<dyn ScalarIndex>>>,
+    vector_cache: Arc<Cache<String, Arc<dyn VectorIndex>>>,
```

Comment:
Do both `ScalarIndex` and `VectorIndx`  share `Index` trait?  Should we consolidate the cache together? might be easier to reason LRU policy.

Reply:
Yes, they both share the `Index` trait.  However, I cannot find any great way to go from `Arc<dyn Index>` to `Arc<dyn ScalarIndex>`.  The best I could find was to add these methods to `Index` trait:

```
fn as_vector(Arc<Self>) -> Option<Arc<dyn ScalarIndex>>;
fn as_scalar(Arc<Self>) -> Option<Arc<dyn VectorIndex>>;
```

That means every index has to implement both `as_vector` and `as_scalar` though which is not very hard but feels a little weird.